lib/mllparty/connection_hub/client_monitor.ex

```diff
@@ -11,91 +10,129 @@ defmodule MLLParty.ConnectionHub.ClientMonitor do
 
   alias MLLParty.ConnectionHub.ClientWrapper
 
-  @impl true
-  def init(_opts), do: {:ok, %{
-    down_at: nil,
-    down_notification_sent: false
-  }}
+  def child_spec(ip, port, client_wrapper_pid) do
+    %{
+      id: "#{ip}:#{port} monitor",
+      start: {__MODULE__, :start_link, [ip, port, client_wrapper_pid]}
+    }
+  end
 
-  def start_link() do
-    GenServer.start_link(__MODULE__, [], name: @name)
+  @impl true
+  def init([ip, port, client_wrapper_pid]) do
+    Logger.info("Starting client monitor: #{ip}:#{port}")
+    schedule_monitor()
+    # Return :ok and initial state
+    {:ok, %{
+      ip: ip,
+      port: port,
+      client_wrapper_pid: client_wrapper_pid,
+      down_at: nil,
+      down_notification_sent: false,
+      reconnect_attempted: false
+    }}
   end
 
-  def start_monitor(client_wrapper_pid) do
-    %{endpoint: endpoint} = ClientWrapper.client_status(client_wrapper_pid)
-    Logger.info("Monitoring connection for #{endpoint}")
-    {:ok, pid} = start_link()
-    Process.send(pid, {:monitor, client_wrapper_pid}, [])
+  def start_link(ip, port, client_wrapper_pid) do
+    GenServer.start_link(__MODULE__, [ip, port, client_wrapper_pid], name: String.to_atom("#{ip}:#{port} monitor"))
```

Comment:
Calling `String.to_atom(...)` for the process name is sensible but is a "gotcha"– atoms aren't garbage collected and the VM can only support a finite number of unique atoms, so in theory someone could iterate over `port <- 1...n` and crash the service. Probably not something we'd realistically run into, but when a dynamic process name is needed, using a [`:via` tuple](https://hexdocs.pm/elixir/main/Registry.html#module-using-in-via) is best.

Here's a good blogpost from Thoughtbot on the subject: https://thoughtbot.com/blog/how-to-start-processes-with-dynamic-names-in-elixir

Reply:
Good to know!